packages/playground/devEnv/build.ts

```diff
@@ -3,294 +3,224 @@ import {writeFile, readFile} from 'fs/promises'
+          React.createElement(PlaygroundPage, {
+            groups: mapValues(groups, (group) => Object.keys(group)),
+          }),
         ),
-      ),
-    ])
-  })
-  .catch((err) => {
-    console.log(err)
-    return process.exit(1)
-  })
-  .then(async () => {
-    // Only start dev server in dev, otherwise just run build and that's it
-    if (!dev) {
-      return
+      )
+      const styleTags = sheet.getStyleTags() // or sheet.getStyleElement();
+      sheet.seal()
+      return {
+        head: styleTags,
+        html,
+      }
+    } catch (error) {
+      // handle error
+      console.error(error)
+      sheet.seal()
+      process.exit(1)
     }
+  })()
 
-    // We start ESBuild serve with no build config because it doesn't need to build
-    // anything, we are already using ESBuild watch.
-    const {port: esbuildPort} = await esbuild.serve(
-      {servedir: playgroundDir('build')},
-      {},
-    )
+  const _initialBuild = timer('esbuild initial playground entry point builds')
+
+  const esbuildConfig: BuildOptions = {
+    entryPoints,
+    bundle: true,
+    sourcemap: true,
+    outdir: buildDir,
+    target: ['firefox88'],
+    loader: {
+      '.png': 'file',
+      '.glb': 'file',
+      '.gltf': 'file',
+      '.svg': 'dataurl',
+    },
+    define: {
+      ...definedGlobals,
+      'window.__IS_VISUAL_REGRESSION_TESTING': 'true',
+    },
+    banner: liveReload?.esbuildBanner,
+    watch: liveReload?.esbuildWatch && {
+      onRebuild(error, result) {
+        esbuildWatchStop = result?.stop ?? esbuildWatchStop
+        liveReload?.esbuildWatch.onRebuild?.(error, result)
+      },
+    },
+  }
 
-    const proxyServer = createServer((req, res) => {
-      const {url, method, headers} = req
-      if (liveReload?.handleRequest(req, res)) {
-        return
-      }
+  let esbuildWatchStop: undefined | (() => void)
 
-      // Otherwise forward requests to ESBuild server
-      req.pipe(
-        request(
-          {
-            hostname: '0.0.0.0',
-            port: esbuildPort,
-            path: url,
-            method,
-            headers,
-          },
-          (prxRes) => {
-            res.writeHead(prxRes.statusCode!, prxRes.headers)
-            prxRes.pipe(res, {end: true})
-          },
+  await esbuild
+    .build(esbuildConfig)
+    .finally(() => _initialBuild.stop())
+    .catch((err) => {
+      // if in dev mode, permit continuing to watch even if there was an error
+      return options.dev ? Promise.resolve() : Promise.reject(err)
+    })
+    .then(async (buildResult) => {
+      esbuildWatchStop = buildResult?.stop
+      // Read index.html template
+      const index = await readFile(path.join(__dirname, 'index.html'), 'utf8')
+      await Promise.all([
+        // Write home page
+        writeFile(
+          path.join(buildDir, 'index.html'),
+          index
+            .replace(/<\/head>/, `${homeHtml.head}<\/head>`)
+            .replace(/<body>/, `<body>${homeHtml.html}`),
```

Comment:
This keeps the %ENTRYPOINT% placeholder intact, resulting in an error 400. I deliberately replaced the whole body, since we don't want to put anything else there anyway.

Reply:
Thanks for pointing this out!